libs/fluid-elements/virtual-scroll-container/src/lib/virtual-scroll-state.ts

```diff
@@ -0,0 +1,291 @@
+  ): void {
+    let index = -1;
+    let first = -1;
+
+    for (let i = 0; i < itemCount; i++) {
+      // Next item's index
+      index = startAt + i + 1;
+
+      // When rendering new items at the start of the list when the user is
+      // scrolling back up, the index might be negative, depending on the virtual
+      // scroll buffer size and current scroll position
+      // In this case, continue looping to render all the items of the scroll buffer
+      // that are not rendered yet
+      if (index < 0) {
+        continue;
+      }
+
+      // Stop rendering if there are no more items
+      if (index >= this._itemsCount) {
+        break;
+      }
+
+      // Push the new item to the list of rendered items
+      if (first === -1) {
+        first = index;
+      }
+
+      // Store index of first rendered item
+      if (index < this._firstRenderedItemIndex || index === 0) {
+        this._firstRenderedItemIndex = index;
+      }
+
+      // Store index of last rendered item
+      if (index > this._lastRenderedItemIndex) {
+        this._lastRenderedItemIndex = index;
+      }
+    }
+
+    if (
+      this.renderedItemsRange.first === -1 ||
+      startAt < this.renderedItemsRange.first
+    ) {
+      this.renderedItemsRange.first = Math.max(0, startAt);
+    }
+    this.renderedItemsRange.last = this._lastRenderedItemIndex;
+  }
+
+  /** Calculates the indices of the first and last item to be observed */
+  private _calculateObservedItemsIndices(): void {
+    if (this._itemsCount <= this._renderedItemsCount()) {
+      // No need to observe anything since there are not enough items to scroll virtually
+      this.bufferStartObservedItemIndex = -1;
+      this.bufferEndObservedItemIndex = -1;
+    } else if (this._lastRenderedItemIndex === this._itemsCount - 1) {
+      // Last item of the list is rendered, no need to observe the end of the list
+      // Observe the start of the list to adjust the buffer when scrolling towards the start
+      this.bufferEndObservedItemIndex = -1;
+      this.bufferStartObservedItemIndex =
+        this._lastRenderedItemIndex -
+        this._initialBatches * this._itemsBatchSize -
+        Math.floor(this._itemsBufferSize * 0.5);
+    } else if (this.startItemsBufferSize < this._itemsBufferSize) {
+      // No more items to add at the beginning of the list, no need to observe
+      // Observe the end of the list to adjust the buffer when scrolling towards the end
+      this.bufferStartObservedItemIndex = -1;
+      this.bufferEndObservedItemIndex =
+        this._lastRenderedItemIndex - Math.floor(this._itemsBufferSize * 0.5);
+    } else {
+      // Observe start and end of the rendered items
+      this.bufferEndObservedItemIndex =
+        this._lastRenderedItemIndex - Math.floor(this._itemsBufferSize * 0.5);
+      this.bufferStartObservedItemIndex =
+        this.bufferEndObservedItemIndex -
+        this._initialBatches * this._itemsBatchSize -
+        this._itemsBufferSize;
+    }
```

Comment:
please use early exits for the else if madness. And move into pure function that returns an object with the indexes for easier testing later on.